woocommerce/woocommerce-ios
# Blaze: Add picker for target locations

Part of #11681 


## Description

This PR adds a new screen for searching and selecting target locations for Blaze campaign creation.

Changes include:
- Added a computed variable `allParentLocationNames` for `BlazeTargetLocation` for getting all the names of a location's parents.
- Updated campaign creation form and view model to navigate to the location picker view and update the location field upon completion.
- Added new view and view model for location picker.
- Updated `MultiSelectionList` with more options for custom section header/footer and selection behavior.

Other changes:
- Minor update to the budget setting screen to move the total budget field position following p1704876472006789-slack-C03L1NF1EA3.
- `BlazeTargetDevicePickerView`, `BlazeTargetLanguagePickerView`, `BlazeTargetTopicPickerView`: Move `navigationViewStyle` and `tasks` modifiers to the `NavigationView`.


📝 Notes:
- Due to the size of this PR, unit tests for `BlazeTargetLocationPickerViewModel` will be added in a separate PR.
- There's no design for this screen, I came up with it myself 😛 Design review thread here: p1705297994011459-slack-C03L1NF1EA3.

## Testing instructions

- Log in to a store with at least one existing product and is eligible for Blaze.
- Navigate to the Products tab and select any public product.
- On the product form, select Promote with Blaze.
- Notice that the campaign creation form is presented. Select the Location field. Confirm the following: 
   - If no specific location is selected previously, only Everywhere option is available on the Location picker.
   - Tap the search bar, the search view should be displayed. The screen should display the correct state: empty state, loading state, no result, or result list based on whether any query has been entered, if searching is in progress or if there's any search result.
   - Select a search result, the search mode should be dismissed. The selected item should be displayed on the picker view and pre-selected.
   - Tap any item other than Everywhere, the Everywhere row and the item should both be unselected.
   - Select all rows in the Specific Locations selection, the Everywhere row should not  be selected automatically. 
   - Select the Everywhere row, all items should be selected.

## Screenshots


https://github.com/woocommerce/woocommerce-ios/assets/5533851/9adddbbc-0738-433f-a88b-b744b41d8c45

---
